### ISSUE TITLE
$options['limit'] breaking find() query.

### ISSUE THREAD

**Issue by Zelfapp** [mongodb/mongo-php-library#153]
This one has me baffled. We've spent about 6 hours today trying to understand what we're doing wrong when using limit.

**The following code works and returns 4 documents. For this particularly find there are only 4 documents that would ever match:**

```
$filter = [
	'cancel_reason' => null,
	'financial_status' => ['$ne' => 'pending'],
	'tags' => new MongoDB\BSON\Regex("kiosk", 'i'),
	'created_at' => ['$gte' => '2016-03-01T03:00:00-04:00', '$lte' => '2016-04-01T02:59:59-04:00'],
	'$or' => [
	    [
		'tags' => new MongoDB\BSON\Regex("Fort Belvoir VA", 'i')
	    ]
	]
];

$options['skip'] = 0;
$options['limit'] = 4;

$cursor = $collection->find($filter, $options);
```

**If I change limit to 3 like below, no documents are returned the script hangs and then dies with zero php errors and no exceptions being thrown:**

`$options['limit'] = 3;`

**Comment by jmikola** [comment]
@Zelfapp: Can you reproduce this using the extension directly (e.g. [`Query::__construct()`](http://php.net/manual/en/mongodb-driver-query.construct.php)). Also, if you could share the relevant `phpinfo()` for the extension and some sample data for the collection (perhaps just five documents, including the four that would match and one additional document), we can investigate on our end.

**Comment by Zelfapp** [comment]
Here's documents for testing. 4 will match query. I'll test with Query and post results. Below documents you'll find my environment.

```
{
    "created_at" : "2016-03-21T11:18:57-07:00",
    "financial_status" : "paid",
    "cancel_reason" : null,
    "tags" : "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking"
}

{
    "created_at" : "2016-03-19T13:22:10-07:00",
    "financial_status" : "paid",
    "cancel_reason" : null,
    "tags" : "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking"
}

{
    "created_at" : "2016-03-19T07:55:44-07:00",
    "financial_status" : "paid",
    "cancel_reason" : null,
    "tags" : "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking"
}

{
    "created_at" : "2016-03-04T13:08:53-08:00",
    "financial_status" : "paid",
    "cancel_reason" : null,
    "tags" : "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking"
}

{
    "created_at" : "2016-03-01T07:35:50-08:00",
    "financial_status" : "paid",
    "cancel_reason" : null,
    "tags" : "1188, Fort Belvoir  VA, kiosk, Processing-Picking"
}
```

**Environment**
Windows 10
Wamp
PHP 5.5.12
Mongodb driver 1.1.4
Latest version of new php library (Installed with composer: composer require "mongodb/mongodb=^1.0.0")
Mongo instance version 2.4.6

**Comment by jmikola** [comment]
I cannot reproduce this with the following extension test case:

```
--TEST--
PHPLIB-153: $options['limit'] breaking find() query
--SKIPIF--
<?php require __DIR__ . "/../utils/basic-skipif.inc"; ?>
--FILE--
<?php
require_once __DIR__ . "/../utils/basic.inc";

$bulk = new MongoDB\Driver\BulkWrite;
$bulk->insert([
    "created_at" => "2016-03-21T11:18:57-07:00",
    "financial_status" => "paid",
    "cancel_reason" => null,
    "tags" => "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking",
]);
$bulk->insert([
    "created_at" => "2016-03-19T13:22:10-07:00",
    "financial_status" => "paid",
    "cancel_reason" => null,
    "tags" => "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking",
]);
$bulk->insert([
    "created_at" => "2016-03-19T07:55:44-07:00",
    "financial_status" => "paid",
    "cancel_reason" => null,
    "tags" => "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking",
]);
$bulk->insert([
    "created_at" => "2016-03-04T13:08:53-08:00",
    "financial_status" => "paid",
    "cancel_reason" => null,
    "tags" => "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking",
]);
$bulk->insert([
    "created_at" => "2016-03-01T07:35:50-08:00",
    "financial_status" => "paid",
    "cancel_reason" => null,
    "tags" => "1188, Fort Belvoir  VA, kiosk, Processing-Picking",
]);

$manager = new MongoDB\Driver\Manager(STANDALONE);
$manager->executeBulkWrite(NS, $bulk);

$query = new MongoDB\Driver\Query(
    [
        'cancel_reason' => null,
        'financial_status' => ['$ne' => 'pending'],
        'tags' => new MongoDB\BSON\Regex("kiosk", 'i'),
        'created_at' => ['$gte' => '2016-03-01T03:00:00-04:00', '$lte' => '2016-04-01T02:59:59-04:00'],
        '$or' => [
            ['tags' => new MongoDB\BSON\Regex("Fort Belvoir VA", 'i')],
        ],
    ],
    [
        'limit' => 3,
        'skip' => 0,
        'projection' => ['_id' => 0],
    ]
);

$cursor = $manager->executeQuery(NS, $query);

var_dump($cursor->toArray());

?>
===DONE===
<?php exit(0); ?>
--EXPECTF--
array(3) {
  [0]=>
  object(stdClass)#%d (%d) {
    ["created_at"]=>
    string(25) "2016-03-21T11:18:57-07:00"
    ["financial_status"]=>
    string(4) "paid"
    ["cancel_reason"]=>
    NULL
    ["tags"]=>
    string(66) "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking"
  }
  [1]=>
  object(stdClass)#%d (%d) {
    ["created_at"]=>
    string(25) "2016-03-19T13:22:10-07:00"
    ["financial_status"]=>
    string(4) "paid"
    ["cancel_reason"]=>
    NULL
    ["tags"]=>
    string(66) "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking"
  }
  [2]=>
  object(stdClass)#%d (%d) {
    ["created_at"]=>
    string(25) "2016-03-19T07:55:44-07:00"
    ["financial_status"]=>
    string(4) "paid"
    ["cancel_reason"]=>
    NULL
    ["tags"]=>
    string(66) "1188, FORT BELVOIR VA, kiosk, Processing-Kiosk, Processing-Picking"
  }
}
===DONE===
```

Note: I excluded the "_id" field in the projection as 2.4 allows it to be appended at the end of the document, while 2.6+ ensures it is serialized first. That said, I ran this test on minor versions of the database through 3.2 (with and without the "_id" field excluded) without issue -- I merely had to adjust the expected output when testing against 2.4. None of the versions resulted in a hang.

The only thing I can suggest is collecting output with the [`mongodb.debug` INI setting](http://php.net/manual/en/mongodb.configuration.php) enabled. You can use "stderr" if you're running this from the CLI or specify a writeable directory if it's being invoked from a web SAPI.

**Comment by Zelfapp** [comment]
I've attached the mongodb.debug error log PHO45F6.txt. In my CLI you can see it ends in "segmentation fault".

![segmentation-fault](https://cloud.githubusercontent.com/assets/11559805/14220376/a3a5d650-f814-11e5-9816-57ccc85c34d9.png)

[PHO45F6.txt](https://github.com/mongodb/mongo-php-library/files/200373/PHO45F6.txt)

**Comment by jmikola** [comment]
Based on that trace, you're using version 1.1.4 of the library with bundled versions of libmongoc and libbson. 1.1.5 is the latest stable release and includes an important fix for Linux/OSX users. Please upgrade and see if that resolves the segfault.

If not, please follow-up with a [GDB backtrace](https://bugs.php.net/bugs-generating-backtrace.php) from 1.1.5 so we can diagnose this further.

**Comment by Zelfapp** [comment]
I have the latest windows dll installed from here: https://pecl.php.net/package/mongodb/1.1.5/windows. It is running libbson version 1.13, libmongoc version 1.12 as you can see in screenshot below.

![mongodb-php](https://cloud.githubusercontent.com/assets/11559805/14253426/4eb00b60-fa40-11e5-85fb-e1582e99e8bc.png)

Also, we tested against Mongo instance mongod version: 2.6.12 and mongod version: 3.0.11. It is definitely an issue with windows DLL's it seems. We don't have a linux setup to test on unfortunately.

**Comment by Zelfapp** [comment]
@jmikola You might not have seen my last answer. We're a bit stuck. Any more ideas for us?

**Comment by jmikola** [comment]
Unfortunately, that strace output isn't helpful at pinpointing the cause of the segfault, which is why we'd need a proper backtrace to get to the heart of it. Based on the mongodb.debug output you attached earlier, the request looks successful. We see connections initialized, some documents returned via the query, and then resources cleaned up and the socket left persisted for the subsequent request(s).

**Comment by Zelfapp** [comment]
We've tested on several different Windows 10 PC's at our workplace thinking it might be one PC, but all PC's have this seg fault failure. Now what?? We spent about 5 hours trying to get gdb running on blasted windows, but we could not get it to interrupt. There's a lot of other people complaining about that with no solutions that worked for us. No windows environments you guys could test on at Mongo? We're dying for this upgrade or we need to figure a different data store I guess. Would love to provide any more info I can, but getting gdb running on windows right now seems impossible to me having never used it before.

**Comment by jmikola** [comment]
@bjori pointed me to https://bugs.php.net/bugs-generating-backtrace-win32.php, which has instructions for generating backtraces on Windows systems (with and without compilers). Is that more helpful?

I extrapolated the test script above into a stand-alone PHP file: https://gist.github.com/jmikola/240af945eb396c8da46d2224571e7d98. You'll need to edit the defined constants atop the file to customize the server, database, and collection.

I'm currently traveling and won't have access to our Windows build machine (Windows 7) until Monday. Once I'm back, I can test the above script as well. We currently do not have a Windows 10 environment for testing the driver.

**Comment by Zelfapp** [comment]
This is not an issue with mongodb on windows at all it turns out! 

We're using the [Codeigniter 3.0.6](https://codeigniter.com/) framework. I shared a test repo with you that has a test controller running inside the CI framework and then the mongo file you provided running outside the framework. The simple test controller running your code breaks when options limit is 3, but outside of CI framework it works every time.

I know having you look at it goes way above and beyond what I would expect a developer to do, but if you have a moment and could debug it would be greatly appreciated. I'm going to share it with the CI developers as well.

Thanks for your help. I'm still trying to get the debug trace working as well.

**Comment by pix0id** [comment]
Hello,

Using the repo Zelfapp posted, I performed the same test on a mac, and am getting the same results.

**Comment by jmikola** [comment]
@pix0id: If you're able to reproduce the segfault on a Mac, can you supply a backtrace?

@Zelfapp: Thanks for the application repo. I'm install the PHP 5.5 version of WAMP server on Windows 7 to test this now.

**Comment by jmikola** [comment]
@Zelfapp: I believe I'm able to reproduce the same error with your application repository. I installed WAMP with PHP 5.5, MongoDB 2.4.14. and the same extension and library versions you cited above.

While I don't see any evidence of a segfault in PHP logs, the Apache error log does report httpd.exe being restarted with a new worker pool. There is also evidence of a crash in Windows' event logs (access through the "Computer Management" app in Control Panel).

Through some trial and error, this appears to be related to the Cursor object being garbage collected. I was able to avoid the segfault by manually unsetting the `$cursor` object in your controller function after iteration (i.e. `$cursor->toArray()`). If that works for you, I would suggest it until we can get around to properly diagnosing and fixing this issue.

**Comment by Zelfapp** [comment]
`unset($cursor);` does the trick.

Really appreciate you digging into this and identifying a workaround.

**Comment by jmikola** [comment]
Thankfully, I was able to generate a backtrace using PHP's built-in web server on Linux. Please follow [PHPC-671](https://jira.mongodb.org/browse/PHPC-671) for further updates. I'll swing back to this thread once a fix is on the way.

**Comment by jmikola** [comment]
This has been fixed in `ext-mongodb` [1.1.7](https://github.com/mongodb/mongo-php-driver/releases/tag/1.1.7).
